src/compiler/mir/typetable.rs

```diff
@@ -0,0 +1,584 @@
+            MirTypeDef::Base(MirBaseType::I8),
+            MirTypeDef::Base(MirBaseType::I16),
+            MirTypeDef::Base(MirBaseType::I32),
+            MirTypeDef::Base(MirBaseType::I64),
+            MirTypeDef::Base(MirBaseType::F64),
+            MirTypeDef::Base(MirBaseType::Bool),
+            MirTypeDef::Base(MirBaseType::StringLiteral),
+        ];
+
+        TypeTable { table }
+    }
+
+    pub fn get(&self, id: TypeId) -> &MirTypeDef {
+        &self.table[id.0 as usize]
+    }
+
+    fn get_mut(&mut self, id: TypeId) -> &mut MirTypeDef {
+        &mut self.table[id.0 as usize]
+    }
+
+    /// Adds the given [`Type`] to the type table. If this type references any type which is
+    /// not in the table then it will also add the referenced type to the table. Structures will
+    /// be added as [`MirStructDef::Declared`].
+    pub fn add(&mut self, ty: &Type) -> TypeId {
+        // Check if ty is already in the table
+        if let Some(id) = self.find(ty) {
+            return id;
+        }
+
+        // Create a MirType using the referenced TypeIds and add to the table
+        let mir_ty = match ty {
+            Type::Array(el, sz) => MirTypeDef::Array {
+                ty: self.add(el),
+                sz: *sz,
+            },
+            Type::RawPointer(mutable, target) => MirTypeDef::RawPointer {
+                mutable: *mutable,
+                target: self.add(target),
+            },
+            Type::Custom(path) => {
+                if path.is_canonical() {
+                    MirTypeDef::Structure {
+                        path: path.clone(),
+                        def: MirStructDef::Declared,
+                    }
+                } else {
+                    panic!("Exected canonical path on custom type")
+                }
+            }
+            _ => panic!("Base types must be in the type table before any other type is added"),
+        };
+
+        self.table.push(mir_ty);
+
+        // Return the index of the new type as a TypeId
+        TypeId(self.table.len() as u32 - 1)
+    }
+
+    /// Adds the definition for the given structure. This will use the path in the
+    /// [`SemanticContext`] as the canonical path for the structure.
+    pub fn add_struct_def(&mut self, sd: &StructDef<SemanticContext>) -> TypeId {
+        // Convert the fields of the structure to MIR Fields
+        let fields = sd.get_fields().iter().map(|f| self.to_field(f)).collect();
+
+        // Search the table for a structure with the same canonical path
+        if let Some(id) = self.find_by_path(sd.context().canonical_path()) {
+            if let MirTypeDef::Structure { def, .. } = self.get_mut(id) {
+                // If a match is found
+                // check if it is not defined
+                if *def == MirStructDef::Declared {
+                    *def = MirStructDef::Defined(fields);
+                    id
+                } else {
+                    panic!("Structure already defined")
```

Comment:
Return `Result` rather than panicking. The user of the `TypeTable` can decide whether it will panic or recover.